crd-generator/api/src/main/java/io/fabric8/crd/generator/CRDGenerator.java

```diff
@@ -0,0 +1,138 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.crd.generator;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator;
+import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator.Feature;
+import io.fabric8.crd.generator.v1.CustomResourceHandler;
+import io.fabric8.kubernetes.api.model.KubernetesList;
+import io.fabric8.kubernetes.client.utils.ApiVersionUtil;
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URI;
+
+public class CRDGenerator {
+
+  private final Resources resources = new Resources();
+  private final CustomResourceHandler v1Handler = new CustomResourceHandler(resources);
+  private final io.fabric8.crd.generator.v1beta1.CustomResourceHandler v1beta1Handler = new io.fabric8.crd.generator.v1beta1.CustomResourceHandler(
+    resources);
+  private CRDOutput output;
+
+  private static final ObjectMapper YAML_MAPPER = new ObjectMapper(
+    new YAMLFactory()
+      .enable(Feature.MINIMIZE_QUOTES)
+      .enable(Feature.ALWAYS_QUOTE_NUMBERS_AS_STRINGS)
+      .disable(YAMLGenerator.Feature.WRITE_DOC_START_MARKER)
+  );
+
+  static {
+    YAML_MAPPER.configure(SerializationFeature.INDENT_OUTPUT, true);
+    YAML_MAPPER.configure(SerializationFeature.WRITE_NULL_MAP_VALUES, false);
+    YAML_MAPPER.configure(SerializationFeature.WRITE_EMPTY_JSON_ARRAYS, false);
+  }
+
+  public CRDGenerator() {
+  }
+
+  public CRDGenerator inOutputDir(File outputDir) {
+    output = new DirCRDOutput(outputDir);
+    return this;
+  }
+
+  public CRDGenerator withOutput(CRDOutput output) {
+    this.output = output;
+    return this;
+  }
+  
+  public CRDGenerator customResources(CustomResourceInfo... infos) {
+    for (CustomResourceInfo info : infos) {
+      System.out.println(
+        "Generating '"
+          + info.crdName()
+          + "' version '"
+          + info.version()
+          + "' with "
+          + info.crClassName()
+          + " (spec: "
+          + info.specClassName()
+          + " / status: "
+          + info.statusClassName()
+          + ")");
+
+      v1Handler.handle(info);
+      v1beta1Handler.handle(info);
+    }
+    return this;
+  }
+
+  public void generate() {
+    final KubernetesList list = resources.generate();
+    list.getItems().forEach(crd -> {
+      final String version = ApiVersionUtil.trimVersion(crd.getApiVersion());
+      try {
+        try (final OutputStream outputStream = output.outputFor(crd.getMetadata().getName() + "-" + version + ".yml")){
```

Comment:
Maybe I missed it, but I think we don't have any sort of validation for required parameters/configurations (output + list of resources).
In case we don't, can we add some validations for these variables? e.g. `Output directory has not been configured and is required use ".inOutputDir(yourTargetDirectory)" to set the value` or something along those lines.

Reply:
Done.